### PR TITLE
GH-41307: [Java] Use org.apache:apache parent pom version 31

### PR DESCRIPTION
**Pull request by laurentgo**
Use/update Maven modules to `org.apache:apache:31` and clean up Maven modules to remove unnecessary configuration or outdated workarounds

* Add `org.apache:apache:31` to `org.apache.arrow:arrow-bom` and `org.apache.arrow.maven.plugins:arrow-maven-plugins` to make them conformant with ASF standards
* Update `org.apache.arrow:arrow-java-root` parent to `org.apache:parent:31`
* Use `version.*` and other properties to override plugin versions defined by `org.apache:parent`
* Move standalone plugin versions under pluginManagement at the top level
* Cleanup redundant plugin version or configuration declaration
* Update `maven-dependency-plugin` to 3.6.1 and add the required overrides when necessary
* Update `maven-shade-plugin` to 3.5.1 (via `org.apache:parent`)
  - disable reduced dependency pom creation for non-terminal modules
* Remove enforcer check for java and maven version (handled by `org.apache:parent`)
* Remove unnecessary `mvnrepository` link comments
* Remove `m2e.version` property check in profiles (only needed for errorprone plugin configuration which is incompatible with M2E)
* Cleanup `argLine` overrides for surefire/failsafe plugins
* Remove unnecessary `../pom.xml` `<relativePath>` directives
* Remove source/target/encoding configuration properties for `maven-compiler-plugin`, `maven-javadoc-plugin` and `maven-resources-plugin` as it is handled by `org.apache:parent` and plugin themselves
* Remove unnecessary copy of codegen templates in `arrow-vector` module
* Remove unnecessary junit jupiter engine dependencies for surefire/failsafe plugins.
* GitHub Issue: apache/arrow#41307
* GitHub Issue: apache/arrow#41307
* GitHub Issue: #41307

### CODE REVIEW

**Comment by laurentgo** [comment]
@lidavidm re-uploading the pull request with the fix for `maven-shade-plugin` behavior.  Are you aware of any other issue caused by this change reported since the migration? (I checked the current issues but haven't seen anything)

**Comment by laurentgo** [comment]
@github-actions crossbow submit java test-conda-python-3.10-spark-v3.5.0

**Comment by laurentgo** [comment]
The `Java JNI / AMD64 Conda Java C Data Interface Integration (pull_request)` issue seems to be related to #41725

**Comment by vibhatha** [comment]
> Yes, noticed that some flight tests seem flaky? I also found occurrence of this issue in other pull requests like this job for example: https://github.com/apache/arrow/actions/runs/9185953424/job/25260768750

Yes indeed.  Would you mind filing an issue for that?

**Comment by laurentgo** [comment]
@vibhatha it looks like I may need to incorporate some of the work you did for #41586 ?

**Comment by laurentgo** [comment]
@vibhatha would appreciate if you could trigger another crossbow run to validate latest changes

**Review by jbonofre — APPROVED** [review]

**Comment by jbonofre** [comment]
It should fix the problem I mentioned in another issue. Thanks !

**Comment by vibhatha** [comment]
@github-actions crossbow submit -g java

**Comment by vibhatha** [comment]
> @vibhatha would appreciate if you could trigger another crossbow run to validate latest changes

I didn't complete it, just tried a few ideas.

**Comment by vibhatha** [comment]
@laurentgo we have an issue in the dataset module: https://github.com/ursacomputing/crossbow/actions/runs/9216288007/job/25356948996#step:6:55162

**Comment by laurentgo** [comment]
Ok, second failure in a row for `java-jars` so can probably eliminate the flakyness hypothesis. Will keep working on a reproduction.

**Comment by vibhatha** [comment]
Thanks @laurentgo

**Comment by laurentgo** [comment]
Update on trying to reproduce the crossbow issue: I tried a local build and setting up my own crossbow queue but the build is successful for both cases 

Attaching a [link](https://github.com/laurentgo/arrow-crossbow/actions/runs/9350373427/job/25735493863) to my crossbow run. linux/arm64 and sccache support have been disabled but the final java build is against macos/arm64.

I'm otherwise open to suggestions

**Comment by lidavidm** [comment]
Let's rebase and try again. Unfortunately, that boot error in tests happens occasionally and we don't have a good way of debugging it.

@vibhatha @danepitkin I think it would be a good improvement to have these jobs collect JVM dump files in the case of failure so we can try to debug this properly.

**Comment by laurentgo** [comment]
I finally managed to reproduce it on [github](https://github.com/laurentgo/arrow-crossbow/actions/runs/9352161921/job/25741526790), but it's 1 time out of 5 or 6 so far. I'll keep running the job regularly (I modified it to capture test output as well), let's see if I can find the root cause

**Comment by laurentgo** [comment]
Here's the output of the jvm dumpstream file after 15 attempts:
```
# Created at 2024-06-05T08:41:26.092
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.

# Created at 2024-06-05T08:41:26.092
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# A fatal error has been detected by the Java Runtime Environment:'.

# Created at 2024-06-05T08:41:26.092
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.

# Created at 2024-06-05T08:41:26.094
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#  SIGSEGV (0xb) at pc=0x000000012900f0b0, pid=12023, tid=25859'.

# Created at 2024-06-05T08:41:26.094
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.

# Created at 2024-06-05T08:41:26.098
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# JRE version: OpenJDK Runtime Environment Temurin-21.0.3+9 (21.0.3+9) (build 21.0.3+9-LTS)'.

# Created at 2024-06-05T08:41:26.098
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# Java VM: OpenJDK 64-Bit Server VM Temurin-21.0.3+9 (21.0.3+9-LTS, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)'.

# Created at 2024-06-05T08:41:26.098
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# Problematic frame:'.

# Created at 2024-06-05T08:41:26.101
Corrupted channel by directly writing to native stream in forked JVM 2. Stream 'C  [jnilib-9883367339098451313.tmp+0x12bb0b0]  arrow::PoolBuffer::Reserve(long long)+0xc4'.

# Created at 2024-06-05T08:41:26.101
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.

# Created at 2024-06-05T08:41:26.102
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again'.

# Created at 2024-06-05T08:41:26.102
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.

# Created at 2024-06-05T08:41:26.104
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# An error report file with more information is saved as:'.

# Created at 2024-06-05T08:41:26.118
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# /Users/runner/work/arrow-crossbow/arrow-crossbow/arrow/java/dataset/hs_err_pid12023.log'.

# Created at 2024-06-05T08:41:26.118
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.

# Created at 2024-06-05T08:41:26.119
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# If you would like to submit a bug report, please visit:'.

# Created at 2024-06-05T08:41:26.121
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#   https://github.com/adoptium/adoptium-support/issues'.

# Created at 2024-06-05T08:41:26.127
Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.
```

**Comment by vibhatha** [comment]
> Let's rebase and try again. Unfortunately, that boot error in tests happens occasionally and we don't have a good way of debugging it.
> 
> @vibhatha @danepitkin I think it would be a good improvement to have these jobs collect JVM dump files in the case of failure so we can try to debug this properly.

Yes, @lidavidm we should. I created an issue https://github.com/apache/arrow/issues/41982

**Comment by vibhatha** [comment]
> Here's the output of the jvm dumpstream file after 15 attempts:
> 
> ```
> # Created at 2024-06-05T08:41:26.092
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.
> 
> # Created at 2024-06-05T08:41:26.092
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# A fatal error has been detected by the Java Runtime Environment:'.
> 
> # Created at 2024-06-05T08:41:26.092
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.
> 
> # Created at 2024-06-05T08:41:26.094
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#  SIGSEGV (0xb) at pc=0x000000012900f0b0, pid=12023, tid=25859'.
> 
> # Created at 2024-06-05T08:41:26.094
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.
> 
> # Created at 2024-06-05T08:41:26.098
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# JRE version: OpenJDK Runtime Environment Temurin-21.0.3+9 (21.0.3+9) (build 21.0.3+9-LTS)'.
> 
> # Created at 2024-06-05T08:41:26.098
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# Java VM: OpenJDK 64-Bit Server VM Temurin-21.0.3+9 (21.0.3+9-LTS, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)'.
> 
> # Created at 2024-06-05T08:41:26.098
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# Problematic frame:'.
> 
> # Created at 2024-06-05T08:41:26.101
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream 'C  [jnilib-9883367339098451313.tmp+0x12bb0b0]  arrow::PoolBuffer::Reserve(long long)+0xc4'.
> 
> # Created at 2024-06-05T08:41:26.101
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.
> 
> # Created at 2024-06-05T08:41:26.102
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again'.
> 
> # Created at 2024-06-05T08:41:26.102
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.
> 
> # Created at 2024-06-05T08:41:26.104
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# An error report file with more information is saved as:'.
> 
> # Created at 2024-06-05T08:41:26.118
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# /Users/runner/work/arrow-crossbow/arrow-crossbow/arrow/java/dataset/hs_err_pid12023.log'.
> 
> # Created at 2024-06-05T08:41:26.118
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.
> 
> # Created at 2024-06-05T08:41:26.119
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '# If you would like to submit a bug report, please visit:'.
> 
> # Created at 2024-06-05T08:41:26.121
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#   https://github.com/adoptium/adoptium-support/issues'.
> 
> # Created at 2024-06-05T08:41:26.127
> Corrupted channel by directly writing to native stream in forked JVM 2. Stream '#'.
> ```

@lidavidm seems like we have an issue here

```bash
arrow::PoolBuffer::Reserve(long long)
```

We probably need to get this dump, download locally and debug?

**Comment by lidavidm** [comment]
Given the small offset I would guess that Reserve was called on a nullptr buffer

**Comment by lidavidm** [comment]
That said I don't think this has to block the PR?

**Comment by vibhatha** [comment]
Yes I think the same.

**Comment by laurentgo** [comment]
I was able to get a reproduction and capture jvm output, hotspot error log and core dump. I put the link to the archive in #32945 .

**Comment by lidavidm** [comment]
@github-actions crossbow submit -g java

File: ci/scripts/java_full_build.sh
Comment (lidavidm) [inline ci/scripts/java_full_build.sh:66]: This is covered by the line below?

File: java/flight/flight-core/pom.xml
Comment (lidavidm) [inline java/flight/flight-core/pom.xml:160]: Hmm, why does this not apply anymore? We saw problems last time we tried to upgrade

**Review by lidavidm — APPROVED** [review]

File: ci/scripts/java_full_build.sh
Comment (laurentgo) [reply ci/scripts/java_full_build.sh:66]: This is covered by the `install` phase which also includes the previous phases like compile or package. The same steps are already declared in the apache parent pom when using the `arrow-release` profile.
- `assembly:single` was used to generate a single artifact with the whole tree and is handled by the parent plugin with the [`source-release-assembly`](https://github.com/apache/maven-apache-parent/blob/36c52ca00b390762f2571fadf4802497b7b80cab/pom.xml#L441) execution in the parent pom (I should also remove the `-DdescriptorId=` property as well I guess)
- `source:jar` is handled by [`attach-sources`](https://github.com/apache/maven-apache-parent/blob/36c52ca00b390762f2571fadf4802497b7b80cab/pom.xml#L462) execution bound by default to `package` phase 
- `javadoc:jar` is handled by the [`attach-javadocs`](https://github.com/apache/maven-apache-parent/blob/36c52ca00b390762f2571fadf4802497b7b80cab/pom.xml#L474) execution bound by default to the `package` phase

File: java/flight/flight-core/pom.xml
Comment (laurentgo) [reply java/flight/flight-core/pom.xml:160]: Yes, [as discussed in the issue](https://github.com/apache/arrow/issues/41571#issuecomment-2123679316) associated to the rollback, the problem was caused by the combination of arrow configuration replaced the main artifact and its associated pom with the shaded artifact and its reduced pom.xml, and the newer versions of the plugins fixing an issue where the reduced pom.xml was not attached to the artifact.

It is solved by disabling the reduced pom.xml generation and attachment using the `createDependencyReducedPom` configuration property

File: ci/scripts/java_full_build.sh
Comment (laurentgo) [reply ci/scripts/java_full_build.sh:66]: Ah sorry, might I misread your question. If you're asking about removing the `find` command, yes, it's because it's handled by the second `find` command which will get the single assembly, the sources and javadoc jars from the local repository and copy them to the `$dist_dir` directory

**Comment by lidavidm** [comment]
@github-actions crossbow submit -g java

**Comment by lidavidm** [comment]
I'll merge once I confirm the crossbow jobs

**Comment by lidavidm** [comment]
Thanks @laurentgo for sticking with this!

**Comment by lidavidm** [comment]
Ah thanks, I was just about to ask about that

**Comment by lidavidm** [comment]
@github-actions crossbow submit java-jars

**Comment by laurentgo** [comment]
> I'll merge once I confirm the crossbow jobs

My bad, we [discussed](https://github.com/apache/arrow/pull/41772#discussion_r1628574206) about the artifacts generated and I forgot to update the task verifying those. My change include a change in `maven/` module to not generate a source release zip archive (containing the whole java codebase) because the main artifact `arrow-java-root-{version}-source-release.zip` already includes the whole tree (maven module included)

**Pull request merged**
